promexporter/targetgroup.go

```diff
@@ -13,40 +13,55 @@
 
 package promexporter
 
-// Target is a prometheus scrape target
-// Labels must be a comparable struct, see `GetLabelNames` function description
-type Target struct {
+import (
+	"fmt"
+
+	"github.com/relex/gotils/generics"
+	"golang.org/x/exp/slices"
+)
+
+// Target is a prometheus scrape target.
+//
+// Labels must be a comparable struct, see `GetLabelNames` function description.
+type Target[L comparable] struct {
 	Target string
-	Labels interface{}
+	Labels L
 }
 
 // TargetGroup is a group of scrape targets sharing the same labels.
+//
 // Each group normally represent the deployment one cluster on one server.
-type TargetGroup struct {
-	Targets []string    `json:"targets"`
-	Labels  interface{} `json:"labels"`
+type TargetGroup[L comparable] struct {
+	Targets []string `json:"targets"`
+	Labels  L        `json:"labels"`
 }
 
 // GroupTargets groups targets by distinct label values
-func GroupTargets(targetList []Target) []TargetGroup {
-	targetGroupByLabels := make(map[interface{}]TargetGroup)
-
-	for _, target := range targetList {
-		if group, exists := targetGroupByLabels[target.Labels]; exists {
-			group.Targets = append(group.Targets, target.Target)
```

Comment:
This was the bug - group.Targets were assigned but the groups in the targetGroupByLabels remained unchanged.

Reply:
do you mean that the first branch of the if clause was never entered?

Reply:
It did enter, but `group` merely holds a copy of the actual value in the `targetGroupByLabels` map, thus replacing group.Targets has no effect outside of that loop.